airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_incremental.py

```diff
@@ -136,6 +142,66 @@ def test_two_sequential_reads(
 
+    def test_read_sequential_slices(
+        self, inputs: IncrementalConfig, connector_config, configured_catalog_for_incremental, cursor_paths, docker_runner: ConnectorRunner
+    ):
+        """
+        Incremental test that makes an initial API request without a state checkpoint. Then we partition the results by stream and
+        slice checkpoints resulting in batches of messages that look like:
+        <state message>
+        <record message>
+        ...
+        <record message>
+
+        Using these batches, we then make API requests using the state message and verify the the correctness of the messages in the
+        response.
+        """
+        if inputs.skip_new_incremental_tests:
+            pytest.skip("Skipping new incremental test based on acceptance-test-config.yml")
+
+        threshold_days = getattr(inputs, "threshold_days") or 0
+        stream_mapping = {stream.stream.name: stream for stream in configured_catalog_for_incremental.streams}
+
+        output = docker_runner.call_read(connector_config, configured_catalog_for_incremental)
+        records_1 = filter_output(output, type_=Type.RECORD)
+        states_1 = filter_output(output, type_=Type.STATE)
+
+        assert states_1, "Should produce at least one state"
+        assert records_1, "Should produce at least one record"
+
+        latest_state = states_1[-1].state.data
+        for record_value, state_value, stream_name in records_with_state(records_1, latest_state, stream_mapping, cursor_paths):
+            assert (
+                record_value <= state_value
+            ), f"First incremental sync should produce records younger or equal to cursor value from the state. Stream: {stream_name}"
+
+        # Create partitions made up of one state message followed by any records that come before the next state
+        filtered_messages = [message for message in output if message.type == Type.STATE or message.type == Type.RECORD]
+        right_index = len(filtered_messages)
+        checkpoint_messages = []
+        for index, message in reversed(list(enumerate(filtered_messages))):
+            if message.type == Type.STATE:
+                checkpoint_messages.insert(0, filtered_messages[index:right_index])
+                right_index = index
+
+        # We sometimes have duplicate identical state messages in a stream which we can filter out to speed things up
+        checkpoint_messages = [message for index, message in enumerate(checkpoint_messages) if message not in checkpoint_messages[:index]]
+
+        # To avoid spamming APIs we only test a fraction of slices
+        sample_rate = 1 if len(checkpoint_messages) <= 5 else len(checkpoint_messages) // 5
+        for message_batch in checkpoint_messages[::sample_rate]:
+            assert len(message_batch) > 0 and message_batch[0].type == Type.STATE
+            current_state = message_batch.pop(0)
```

Comment:
it may seem a bit heavy handed to assemble all these partitions just to pop off the state which is used in the next `call_read_with_state`, but its meant to set up the next part validation in a subsequent PR which uses the record messages from the batch and verifies that they are all present in the result of the current `call_read_with_state`. 

Alternatively, I can remove all the logic and just build a list of state messages, if the check-ins should remain isolated